Fragment shader barycentric extension

Szymon-Rajczakowski opened the pull request:
## Description

Demonstrate how to use fragment shader barycentric feature, which allows accessing barycentric coordinates for each processed fragment.

### Notes

Sample was tested using following configurations:

Windows 11 Education 22H2
NVIDIA GeForce RTX 3080 Ti (NVIDIA driver version: 31.0.15.3118, driver date: 24.02.2023)
Intel(R) Core(TM) i7-9700K
16GB RAM

Windows 10 Pro
Nvidia T600 Laptop GPU (NVIDIA driver version: 31.0.15.2849, driver date: 02.02.2023)
Intel i7-11850H
32GB RAM



## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

Commit:
Create draft of fragment_shader_barycentric sample

Commit:
Preparing graphic pipeline
Enabling fragment shader barycentric extension

Commit:
- fix problem with enabling
fragment_shader_barycentric extension
- Add draft shaders with barycentric coordinates

Commit:
update glslang submodule for
enabling fragment shader barycentric extesnion

Commit:
Dividing gbuffer shader to object and skybox shader
Adding switch case to object frag shader
Adding push constants
Adding on_update_ui_overlay

Karina-Murawko-Wisniewska-Mobica committed:
Update sample and add readme draft

Commit:
Change model name from "model" to "object"
Fix problem with cube texture
README refactor

Karina-Murawko-Wisniewska-Mobica committed:
Add texture effect

Karina-Murawko-Wisniewska-Mobica committed:
Update README

Karina-Murawko-Wisniewska-Mobica committed:
Fix spelling

Commit:
Change name of effects
Spelling fix

Commit:
Remove unnecessary descriptor_pool.

Commit:
Add image

Commit:
Merge branch 'main' of https://github.com/Oskar-Plaska-Mobica/Vulkan-Samples into fragment_shader_barycentric_extension

Szymon-Rajczakowski committed:
Fix Copyright Header

gpx1000 commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,429 @@
+#include "common/vk_common.h"
+#include "gltf_loader.h"
+#include "gui.h"
+#include "platform/filesystem.h"
```
don't think you need filesystem included.

Szymon-Rajczakowski replied:
fixed

gpx1000 commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,429 @@
+#include "platform/filesystem.h"
+#include "platform/platform.h"
+#include "rendering/subpasses/forward_subpass.h"
+#include "stats/stats.h"
```
stats.h doesn't look like it is used.

Szymon-Rajczakowski replied:
fixed

gpx1000 commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,429 @@
+	shader_stages[1] = load_shader("fragment_shader_barycentric/skybox.frag", VK_SHADER_STAGE_FRAGMENT_BIT);
+
+	// Create graphics pipeline for dynamic rendering
+	VkFormat color_rendering_format = render_context->get_format();
```
I don't see where color_rendering_format is used other than to just set it here.

Szymon-Rajczakowski replied:
fixed

SaschaWillems reviewed (dismissed):
LGTM and runs fine on my system (NV RTX 2060, Win10)

Commit:
Remove unnecessary lines

Szymon-Rajczakowski committed:
Merge branch 'main' into fragment_shader_barycentric_extension

asuessenbach commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,418 @@
+		render_pass_begin_info.framebuffer              = framebuffers[i];
+		render_pass_begin_info.renderArea.extent.width  = width;
+		render_pass_begin_info.renderArea.extent.height = height;
+		render_pass_begin_info.clearValueCount          = 3;
```
`3` -> `static_cast<uint32_t>(clear_values.size())`
Note: there are just 2 clear_values!

Szymon-Rajczakowski replied:
fixed, thanks for review!

asuessenbach commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,418 @@
+	        1,
+	        &blend_attachment_state);
+	color_blend_state.attachmentCount = 1;
+	color_blend_state.pAttachments    = &color_attachment_state;
```
`color_attachment_state` overwrites `blend_attachment_state` as attachment.
One of them can be removed.

Szymon-Rajczakowski replied:
fixed, thanks for review!

asuessenbach commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,418 @@
+		range.layerCount     = VK_REMAINING_ARRAY_LAYERS;
+
+		VkImageSubresourceRange depth_range{range};
+		depth_range.aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
```
`range` is just used by `depth_range`, and `depth_range` is never used.
They both can be removed.

Szymon-Rajczakowski replied:
fixed, thanks for review!

asuessenbach commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.h:
```diff
@@ -0,0 +1,88 @@
+		                                                  "DIFF_PERSP_NOPERSP",
+		                                                  "WIREFRAME",
+		                                                  "INTERP_TO_MASS_CENTER",
+		                                                  "BARYCOORD_TEXTURE"};
```
As these are strings in the UI, you might use some text that is better human-readable? Like "color interpolation", "perspective vs. non-perspective",...

Szymon-Rajczakowski replied:
fixed, thanks for review!

asuessenbach reviewed (commented):
Great sample!
Just three minor and one substantial issue.

Commit:
- Fix bug with wrong size of clear value
- Remove unnecessary lines
- Change strings in the UI to more human-readable
- Update README

asuessenbach commented on samples/extensions/fragment_shader_barycentric/fragment_shader_barycentric.cpp:
```diff
@@ -0,0 +1,404 @@
+		render_pass_begin_info.framebuffer              = framebuffers[i];
+		render_pass_begin_info.renderArea.extent.width  = width;
+		render_pass_begin_info.renderArea.extent.height = height;
+		render_pass_begin_info.clearValueCount          = 2;
```
`2` -> `static_cast<uint32_t>(clear_values.size())` ?

Szymon-Rajczakowski replied:
That is definitely a better option than hardcoded value. This will help avoid mistakes like the one I made before (where I inserted 3 instead of 2). Thanks for your review!

Szymon-Rajczakowski committed:
Update fragment_shader_barycentric.cpp

use static_cast<uint32_t>(clear_values.size()) instead of hardcoding.
